pyfr/backends/cuda/cublaslt.py

```diff
@@ -0,0 +1,296 @@
+        self._create(ptr, dtype, m.ncol, m.nrow, m.leaddim)
+        self._as_parameter_ = ptr.value
+
+
+GEMMDesc = namedtuple('GEMMDesc', ['mm', 'a', 'b', 'c', 'algo', 'ws_ptr',
+                                   'ws_size'])
+
+
+class CUDACUBLASLtKernels(CUDAKernelProvider):
+    def __init__(self, backend):
+        super().__init__(backend)
+
+        self.handle = c_void_p()
+
+        # Load and wrap CUBLASLt
+        self.lib = CUBLASLtWrappers()
+
+        # Init
+        self.lib.cublasLtCreate(self.handle)
+
+        # Maximum number of algorithms to test
+        self.nkerns = backend.cfg.getint('backend-cuda', 'cublas-nkerns', 512)
+
+        # GEMM cache
+        self._mul_cache = {}
+
+    def __del__(self):
+        if self.handle:
+            self.lib.cublasLtDestroy(self.handle)
+
+    def mul(self, a, b, out, alpha=1.0, beta=0.0):
+        cuda = self.backend.cuda
+        w, h = self.lib, self.handle
+
+        # Ensure the matrices are compatible
+        if a.nrow != out.nrow or a.ncol != b.nrow or b.ncol != out.ncol:
+            raise ValueError('Incompatible matrices for out = a*b')
+
+        # cuBLAS expects inputs to be column-major (or Fortran order in
+        # numpy parlance).  However as C = A*B => C^T = (A*B)^T
+        # = (B^T)*(A^T) with a little trickery we can multiply our
+        # row-major matrices directly.
+        m, n, k = b.ncol, a.nrow, a.ncol
+        A, B, C = b, a, out
+
+        # Cache key
+        ckey = (A.dtype, alpha, beta, m, n, k, A.leaddim, B.leaddim, C.leaddim)
+
+        # Size checks
+        if any(sz > 2**31 - 1 for sz in ckey[3:]):
+            raise ValueError('Matrices too large for cuBLAS')
+
+        # α and β factors for D = α*(A*B) + β*C
+        if a.dtype == np.float64:
+            alpha_ct, beta_ct = c_double(alpha), c_double(beta)
+            ctype = w.CUBLAS_COMPUTE_64F
+            dtype = w.CUDA_R_64F
+        else:
+            alpha_ct, beta_ct = c_float(alpha), c_float(beta)
+            ctype = w.CUBLAS_COMPUTE_32F
+            dtype = w.CUDA_R_32F
+
+        desc = GEMMDesc(None, None, None, None, None, None, 0)
+
+        # Convenience wrapper
+        def gemm(stream):
+            w.cublasLtMatmul(
+                h, desc.mm, byref(alpha_ct), A, desc.a, B,
+                desc.b, byref(beta_ct), C, desc.c, C, desc.c,
+                byref(desc.algo), desc.ws_ptr, desc.ws_size, stream
+            )
+
+        # Obtain the performance of the kernel
+        try:
+            desc, dt = self._mul_cache[ckey]
+        except KeyError:
+            # Create matrix layouts
+            a_desc = CUBLASLtMatrixLayout(self, h, A, dtype)
+            b_desc = CUBLASLtMatrixLayout(self, h, B, dtype)
+            c_desc = CUBLASLtMatrixLayout(self, h, C, dtype)
+
+            # Matmul descriptor
+            mm_desc = CUBLASLtMatmulDesc(self, h, ctype, dtype)
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSA] = w.CUBLAS_OP_N
```

Comment:
I'd just pass these to the constructor as fixed arguments rather than making it look like a write-only map.